src/index.js

```diff
@@ -105,11 +116,60 @@ export default class BrowserStackConnector {
         const workerId = await createWorker();
         const worker   = await this._getWorker(workerId);
 
+        const waitForUrlOpened = new Promise((resolve, reject) => {
+            let timeoutId = null;
+
+            const hubHandler = id => {
+                if (id === urlIdentifier) {
+                    this.hub.removeListener('open', hubHandler);
+
+                    clearTimeout(timeoutId);
+                    resolve();
+                }
+            };
+
+            timeoutId = setTimeout(async () => {
+                this.hub.removeListener('open', hubHandler);
+
+                await this.stopBrowser(worker.id);
+
+                reject('Browser starting timeout expired');
+            }, START_BROWSER_TIMEOUT);
+
+            this.hub.addListener('open', hubHandler);
+        });
+
+        await waitForUrlOpened;
+
         this._log(`${browserSettings.name} started. See ${worker.browser_url}`);
 
         return worker;
     }
 
+    async startBrowser (browserSettings, url, { jobName, build } = {}, timeout = null) {
+        let worker  = null;
+        let attempt = 0;
+        let error   = null;
+
+        while (attempt < START_BROWSER_MAX_ATTEMPT && !worker) {
+            try {
+                error  = null;
+                worker = await this._startBrowser(browserSettings, url, { jobName, build }, timeout);
+            }
+            catch (err) {
+                error  = err;
+                worker = null;
```

Comment:
it's not necessary to set `worker = null` here because it's `null` if `this._startBrowser()` failed